SLING-10269 cache the result of isResourceType()

joerghoh opened the pull request.

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1046,37 +1051,57 @@ public String getParentResourceType(final String resourceType) {
+
+             // Check if the result is already available from cache
+             String3Tupel key = new String3Tupel(resource.getResourceType(),resource.getResourceSuperType(), resourceType);
+             if (resourceTypeLookupCache.containsKey(key)) {
```
I would suggest a get and null check. This way we don't query the map twice.

joerghoh replied:
Done

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1046,37 +1051,57 @@ public String getParentResourceType(final String resourceType) {
     public boolean isResourceType(final Resource resource, final String resourceType) {
         boolean result = false;
         if ( resource != null && resourceType != null ) {
-             // Check if the resource is of the given type. This method first checks the
```
This inline comment seems to have been lost in the refactoring.

joerghoh replied:
Added it as method description for `isResourceTypeInternal`

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1117,4 +1142,68 @@ public Resource move(final String srcAbsPath, final String destAbsPath) throws P
+
+
+    // A very simple tupel implementation which can be used as key in any map
+    public class String3Tupel {
```
Nit: tuple, not tupel :-)

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1117,4 +1142,68 @@ public Resource move(final String srcAbsPath, final String destAbsPath) throws P
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
```
I think you can use `java.util.Objects.hashCode` here.

joerghoh replied:
Eclipse created both `hashCode` and `equals`, so I didn't want to bother with it :-)

rombert replied:
Well, the maintenance/review cost will be there, so you can ask Eclipse to do the right thing with recent versions :-)

https://stackoverflow.com/a/51775296/112671

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1117,4 +1142,68 @@ public Resource move(final String srcAbsPath, final String destAbsPath) throws P
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + getEnclosingInstance().hashCode();
```
Is this necessary? A tuple3 does not escape the scope of a single resource resolver, right?

rombert replied:
I am still missing the point of this field, seems it's only used in hashcode/equals - is it really necessary?

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1117,4 +1142,68 @@ public Resource move(final String srcAbsPath, final String destAbsPath) throws P
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj)
```
I think you can use `java.util.Objects.equals` here.

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1117,4 +1142,68 @@ public Resource move(final String srcAbsPath, final String destAbsPath) throws P
+            if (getClass() != obj.getClass())
+                return false;
+            String3Tupel other = (String3Tupel) obj;
+            if (!getEnclosingInstance().equals(other.getEnclosingInstance()))
```
Same as above, is this needed?

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1117,4 +1142,68 @@ public Resource move(final String srcAbsPath, final String destAbsPath) throws P
+            return true;
+        }
+
+        String s1;
```
Any reason to not make these fields private? Also, fields should go before contructors, constructors before instance methods :-)

joerghoh replied:
Changed order (not sure what Eclipse did here ...)

rombert commented on src/test/java/org/apache/sling/resourceresolver/impl/ResourceResolverImplTest.java:
```diff
@@ -521,6 +527,28 @@ public void testCloseWithStackTraceLogging() throws Exception {
         assertFalse(resolver.isResourceType(r, "h:p"));
     }
 
+    @Test public void testIsResourceTypeCached() throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
```
Nit: maybe `throws Exception`?

rombert reviewed (commented):
I did an initial pass on the implementation, overall LGTM, but - as always :-) - I have some minor tweaks I think would be good to include.

There are two major questions outstaning for me:
- is the thread safety enough? We typically don't share resource resolvers between threads, but I think the current contract is "you might get stale data". Unsynchronised access to a HashMap might lead to deadlocks
- are we aware of any inconsistent cache scenarios? I _think_ that all changes done during the current session will be reflected in the current ResourceResolver but maybe it's worth analysing this as well.

kwin commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1046,37 +1051,57 @@ public String getParentResourceType(final String resourceType) {
-                 }
+
+             // Check if the result is already available from cache
+             String3Tupel key = new String3Tupel(resource.getResourceType(),resource.getResourceSuperType(), resourceType);
```
Instead of a generic 3 string tuple, a more specific class like `ResourceTypeInformation` with speaking getter method names are way easier to understand.

joerghoh replied:
I renamed the class. We don't need any getters, as it is just used as a key in the map.

joerghoh reviewed (commented):
@rombert: you raised an interesting point: Changes which are performed via ResourceResolver. While I drop now the cache on every commit, It's actually quite hard to do if someone is doing a change directly via a enclosed JCR session. I don't think that there is a way for sling to detect that. Even the [JcrValueMap](https://github.com/apache/sling-org-apache-sling-jcr-resource/blob/master/src/main/java/org/apache/sling/jcr/resource/internal/JcrValueMap.java) is caching properties of JCR nodes within a ValueMap, and doesn't care if it's in sync with the JCR (and it's not synchronized either).

Besides that: Thanks @bdelacretaz, @rombert , @kwin  for your feedback, everything should be addressed.

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -124,6 +129,7 @@ private ResourceResolverImpl(final ResourceResolverImpl resolver, final Map<Stri
         this.context = new ResourceResolverContext(this, factory.getResourceAccessSecurityTracker());
         this.control = createControl(factory.getResourceProviderTracker(), authInfo, resolver.control.isAdmin());
         this.factory.register(this, control);
+        this.resourceTypeLookupCache = new HashMap<>();
```
This one is still unsynchronized, but see new comment below.

rombert commented on src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java:
```diff
@@ -1046,37 +1053,60 @@ public String getParentResourceType(final String resourceType) {
+
+             // Check if the result is already available from cache
+             ResourceTypeInformation key = new ResourceTypeInformation(resource.getResourceType(),resource.getResourceSuperType(), resourceType);
+             Boolean value = resourceTypeLookupCache.get(key);
```
Thinking about it some more, maybe a `ConcurrentMap` is a better fit? It's more expressive, e.g.

```java
resourceTypeLookupCache.computeIfAbset(key, (key) -> isResourceTypeInternal(resource, resourceType));
```

and also skips the locking of the synchronized map.

If you're happy with the synchronized map, feel free to use that.

joerghoh replied:
If the access is normally happening synchronously, the overhead of locking seems acceptable to me compared to the overhead of a more complex implementation. But unless we have benchmarked it's probably an academic discussion :-)

joerghoh commented:
@rombert: Thanks for your feedback, I adjusted the implementation for `hashcode` as requested (and I wonder why Eclipse needs to be told explicitly to make it right).
I also switched to a ConcurrentHashMap, and the use `computeIfAbsent` makes the code indeed much more expressive (and shorter).

rombert reviewed (approved):
LGTM, thanks!

rombert commented:
@joerghoh - all good know from my POV. Please make sure to _Squash and merge_ when done to keep a single commit for this atomic change.
